valueset.go

```diff
@@ -0,0 +1,327 @@
+			first = true
+		}
+		nsLeaf := rec.Values[1].(neo4j.Node)
+		var nsLeafEntityId string
+		for k := range nsLeaf.Props {
+			if k == ls.EntityIDTerm {
+				nsLeafEntityId = k
+			}
+		}
+		ns.Data[nsLeafEntityId] = NodesetData{
+			ID:         nsLeafEntityId,
+			Labels:     nsLeaf.Labels,
+			Properties: nsLeaf.Props,
+		}
+	}
+	return ns, nil
+}
+
+func buildProps(props map[string]interface{}, vars map[string]interface{}) string {
+	out := strings.Builder{}
+	first := true
+
+	for k, v := range props {
+		if v == nil {
+			continue
+		}
+		if first {
+			out.WriteRune('{')
+			first = false
+		} else {
+			out.WriteRune(',')
+		}
+		out.WriteString(quoteBacktick(k))
+		out.WriteRune(':')
+		out.WriteRune('$')
+		tname := fmt.Sprintf("p%d", len(vars))
+		out.WriteString(tname)
+		_, ok := v.(string)
+		if !ok {
+			vars[tname] = v.([]string)
+		} else {
+			vars[tname] = v.(string)
+		}
+	}
+	if !first {
+		out.WriteRune('}')
+	}
+
+	return out.String()
+}
+
+func Execute(ctx context.Context, tx neo4j.ExplicitTransaction, cfg Config, oldNodeset, newNodeset Nodeset, rootOp string, inserts, updates, deletes []NodesetData) error {
+	switch rootOp {
+	case "insertRoot":
+		vars := make(map[string]interface{})
+		if _, err := tx.Run(ctx, fmt.Sprintf("CREATE (n %s %s)", cfg.MakeLabels(newNodeset.Labels), buildProps(newNodeset.Properties, vars)), vars); err != nil {
+			return err
+		}
+	case "updateRoot":
+		if _, err := tx.Run(ctx, fmt.Sprintf("MATCH (n) WHERE n.entityId = $eid SET n.props=$props, n%s", cfg.MakeLabels(oldNodeset.Labels)), map[string]interface{}{
+			"eid": oldNodeset.ID, "props": oldNodeset.Properties}); err != nil {
+			return err
+		}
+	case "deleteRoot":
+		if _, err := tx.Run(ctx, fmt.Sprintf("MATCH (n {nodeset_id: $eid})-[r:%s]->() DELETE r", oldNodeset.ID), map[string]interface{}{"eid": oldNodeset.ID}); err != nil {
+			return err
+		}
+	}
+	type groupedNodesetData struct {
+		labelExpr string
+		data      []NodesetData
+	}
+	groupByLabel := func(nodesetData []NodesetData) []groupedNodesetData {
+		ret := make([]groupedNodesetData, 0)
+		hm := make(map[string][]NodesetData)
+		for _, nsD := range nodesetData {
+			nsD.Labels = sort.StringSlice(nsD.Labels)
+			labelExpr := cfg.MakeLabels(nsD.Labels)
+			hm[labelExpr] = append(hm[labelExpr], nsD)
+		}
+		for expr, nsData := range hm {
+			ret = append(ret, groupedNodesetData{labelExpr: expr, data: nsData})
+		}
+		return ret
+	}
+	mapNodesetData := func(nodesetData []NodesetData) []map[string]interface{} {
+		ans := make([]map[string]interface{}, len(nodesetData))
+		for ix, item := range nodesetData {
+			ans[ix] = structs.Map(item)
+		}
+		return ans
+	}
+	if len(updates) > 0 {
+		for _, update := range groupByLabel(updates) {
+			unwindData := map[string]interface{}{"nodes": mapNodesetData(update.data)}
+			query := fmt.Sprintf("unwind $nodes as node MATCH (n) WHERE n.`https://lschema.org/entityId` = node.ID SET n=node.Properties, n%s", update.labelExpr)
```

Comment:
match (n:<labels>) where n.`%s` where %s is shortened ls.EntityId. use labels. If there are no labels for nodeset data, use CONCEPT